flax/core/scope.py

```diff
@@ -534,14 +534,11 @@ def name_reserved(self, name: str, col: Optional[str] = None) -> bool:
       col: if a variable, the collection used.
     """
     if name in self.reservations:
-      # with relaxed naming, allow the same name for two variables in
+      # allow the same name for two variables in
       # different collections, otherwise raise error.
-      if config.flax_relaxed_naming:
-        if (self.reservations[name] is None or col is None
-            or self.reservations[name] == col):
-            return True
-      else:
-        return True
+      if (self.reservations[name] is None or col is None
```

Comment:
Since a variable name can now belong to multiple collection, `reservation` should be modified to map the name to its known collections e.g.
```python
reservations: Dict[str, Set[str]]
```
Currently, if you reserve a name for collection A, it seems you can then reserve it multiple times for any other collection.

Reply:
ah good catch